haskell/src/CLI.hs

```diff
@@ -3,80 +3,86 @@
+module CLI (main) where
 
-import Env
+import Env hiding (Parser)
 import qualified Lentille.Gerrit as G
 import Macroscope.Main (runMacroscope)
 import qualified Monocle.Api
 import Monocle.Client (withClient)
 import Monocle.Prelude hiding ((:::))
 import Monocle.Search.Query (parseDateValue)
+import Options.Applicative hiding (header, help, str)
 import Options.Generic
 import qualified Streaming.Prelude as S
 
 ---------------------------------------------------------------
--- API cli
+-- Unified CLI
 ---------------------------------------------------------------
 
-data CliEnv = CliEnv
-  { config :: String,
-    elastic_conn :: String
-  }
-
-cliEnv :: IO CliEnv
-cliEnv =
-  Env.parse (header "monocle-api available environement variables") $
-    CliEnv <$> var str "CONFIG" (help "The Monocle configuration" <> def "/etc/monocle/config.yaml" <> helpDef show)
-      <*> var str "ELASTIC_CONN" (help "The Elasticsearch endpoint" <> def "elastic:9200")
-
-newtype CLIOptions w = CLIOption
-  { port :: w ::: Maybe Int <?> "The API port to listen to, default to 9898"
-  }
-  deriving stock (Generic)
-
-instance ParseRecord (CLIOptions Wrapped) where
-  parseRecord = parseRecordWithModifiers lispCaseModifiers
+-- | usage is an optparse-applicative Parser that provides the final action
+-- See the last example of https://github.com/pcapriotti/optparse-applicative#commands
+usage :: Parser (IO ())
+usage =
+  subparser
+    ( mkCommand "Start the API" "api" usageApi
+        <> mkCommand "Start the Crawlers" "crawler" usageCrawler
+        <> mkCommand "Maintain the database" "janitor" usageJanitor
+        <> mkCommand "Run a single crawler standlone" "lentille" usageLentille
+    )
+  where
+    -- The API entrypoint (no CLI argument).
+    usageApi = pure $ do
+      -- get parameters from the environment
+      (config, elastic, port) <- getFromEnv ((,,) <$> envConf <*> envElastic <*> envApiPort)
+      -- start the API
+      Monocle.Api.run (getInt port) (getURL elastic) config
+
+    -- The Crawler entrypoint (no CLI argument).
+    usageCrawler = pure $ do
+      -- get parameters from the environment
+      (config, url, monitoringPort) <-
+        getFromEnv
+          ( (,,)
+              <$> envConf
+              <*> envPublicUrl
+              <*> envMonitoring
+          )
+      -- start the Crawler
+      withClient url Nothing $ runMacroscope (getInt monitoringPort) config
+
+    -- The standalone lentille entrypoint (re-using optparse-generic record)
+    usageLentille = mainLentille <$> parseRecord
+
+    -- The janitor entrypoint
+    usageJanitor = pure $ putStrLn "NotImplemented"
+
+    -- Helper to create sub command
+    mkCommand doc name parser = command name $ info parser $ progDesc doc
+
+    -- Helpers to get value fromm the env
+    getFromEnv = Env.parse (header "monocle")
+    envConf = var str "CONFIG" (help "The Monocle configuration" <> def "/etc/monocle/config.yaml" <> helpDef show)
```

Comment:
I'm not able to discover the environment variables with for instance "cabal run exe:monocle -- api --help"

Reply:
Good idea, using https://hackage.haskell.org/package/optparse-applicative-0.16.1.0/docs/Options-Applicative.html#v:footer we can add extra doc, and we can generate the help from envparse with: https://hackage.haskell.org/package/envparse-0.4.1/docs/Env.html#v:helpDoc